### ISSUE TITLE
Godot OpenXR can't target Windows Mixed Reality OpenXR runtime because it relies on DirectX

### ISSUE THREAD

**Issue by rcoppy** [GodotVR/godot_openxr#51]
Hi there, just wanted to say you've been doing fantastic work!

Fyi--this OpenXR addon doesn't work with the Windows Mixed Reality platform out of the box:

`ERROR: OpenXRApi::initialiseInstance: OpenXR Runtime does not support OpenGL extension!`

Because WMR only support DirectX 😢 

But! Preliminary reading found that someone as part of Google Summer of Code made an OpenGL to DirectX compatibility layer for Blender's OpenXR loader, and it actually (mostly) worked: 

https://devtalk.blender.org/t/gsoc-2019-vr-support-through-openxr-weekly-reports/7665
https://devtalk.blender.org/t/gsoc-2019-core-support-of-virtual-reality-headsets-through-openxr/7614/15

The project is, however, 2 years old and skimming his last update parts of the workaround seem to have broken because of API changes. 

Nonetheless, if WMR compatibility is a priority, @julianeisel 's foundational work is likely helpful. 

Cheers!

**Comment by BastiaanOlij** [comment]
Yup it's a sad sad thing, I so want to see Godot running on a hololens.

It's mostly a matter of people not having the time to spend on it. Godot can be compiled using Angle so it runs on DirectX but every since Microsoft abandoned Angle it's been a pain in the ..... It requires this in the UWP build so Godot can run on XBox, and incidentally also on Hololens :)

Google has now picked it up and there have been a few trials to get Godot running with that version but I don't know the state of that. In theory if you compile Godot with Angle and make the necessary changes to this plugin it should work.

But someone has to find the time to do it. I will get to it eventually but right now there are other things higher on the priority list.

**Comment by GeorgeS2019** [comment]
@BastiaanOlij 
FYI DirectX to Vulkan: is this relevant to bring Godot support for DirectX (Hololens OpenXR) ??

https://synappsis.wordpress.com/2018/03/22/dxvk/

https://github.com/doitsujin/dxvk/

**Comment by BastiaanOlij** [comment]
Wrong way around I'm afraid @GeorgeS2019 , we need Vulkan to DirectX so we can run Godot 4 on systems that only support DirectX.

**Comment by GeorgeS2019** [comment]
Here is the [link](https://github.com/ttauri-project/ttauri/issues/203#issue-949390540)

Not all features of the Windows 10 compositor is available in OpenGL or Vulkan, specifically scRGB float 16 HDR and transparent windows.

It seems it is possible to use a DirectX swap-chain and use Vulkan-DirectX interopability to get Vulkan to render in a frame buffer and without cost 'copy' the frame buffer to the directX swap-chain (like a c++ move). Care must be taken to somehow copy back the directX swap-chain image to the vulkan frame buffer to allow for scissored rendering.

Here is an excerpt from a forum post I found:

P.S. If you want to extend HDR to work in Vulkan, the consensus among all of the AAA games I have reverse engineered to date is to use a DXGI SwapChain + D3D11 Immediate Context to perform Vk / DXGI interop (copy the Vk swpachain's backbuffer to DXGI for Presentation).

It sounds weird, but if you match up the SwapChain image format and dimensions there is actually no overhead introduced by doing this (under the hood it's just a shared surface regardless which API is touching it). Just make sure you have cleanly released all references to said buffer after Present -- there's one implicitly held by the DWM until VBLANK and the D3D11 Immediate Context may hold one unless you call ClearState (...) and Flush (...) prior to Present. You get finer grained control over pretty much everything (latency, scaling, colorspace, G-Sync, ...) if you use IDXGISwapChain*::Present (...) to present final frames for scanout, so this interop is a blessing in disguise for HDR users.

**Comment by BastiaanOlij** [comment]
Hopefully someone will start to look into this for Godot 4 but I won't have time for it myself. There has been talk about doing a proper DirectX backend for Godot 4 once 4.0 has been released but it will likely require either help from Microsoft or someone with sufficient DirectX experience to become a contributor.

The OpenXR plugin just handles the OpenXR layer, until Godot 4 works in some way, shape or form with DirectX there isn't much we can do for the plugin.

**Comment by mbucchia** [comment]
Hello,

I've written an OpenXR API layer that will let Vulkan apps run on top of Direct3D 12 support on WMR (PCVR):

https://github.com/mbucchia/OpenXR-Vk-D3D12

I've tested it with the Godot 4 Alpha 10 demo scene, and it worked well on my HP Reverb G2. This can be installed system-wide (the API layer will load automatically) or bundled with the engine (need to load the API layer with `xrCreateInstance()`).

This won't help you on HoloLens 2, because the GPU driver itself does not support Vulkan, so texture interop (which is the end of the line) will not help here.

**Comment by GeorgeS2019** [comment]
@mbucchia  In your view, what will be needed to make Godot4 run in HoloLens 2?  Do you use Dozen? Great Job!!

**Comment by mbucchia** [comment]
> @mbucchia In your view, what will be needed to make Godot4 run in HoloLens 2? Do you use Dozen? Great Job!!

AFAIK the GPU driver itself on HoloLens 2 does not support Vulkan/OpenGL. The best solution is for the app to natively support Direct3D. Hacky way would be a wrapper Vulkan -> Direct3D or OpenGL -> Direct3D. As pointed out earlier, dxvk is the other way around, so no luck. There is OpenGLon12 (now part of Mesa I believe) that I looked into a while back, and it would be possible to modify OpenGLon12 to retrieve the underlying Direct3D 12 texture, and copy it the OpenXR swapchain. I have however no idea whether OpenGLon12 has sufficient OpenGL support for Godot.

**Comment by GeorgeS2019** [comment]
@mbucchia  
I believe **Dozen** could be a better choice  
[15 June 2022: Microsoft's "Dozen" Prepares For Vulkan 1.1 Support](https://www.phoronix.com/scan.php?page=news_item&px=Microsoft-Dzn-Vulkan-1.1-Nears)

**Comment by mbucchia** [comment]
I wasn't aware of this, but it's the same concept as OpenGLon12. Looking briefly at the code, it's also possible to modify the implementation to "reach" into internals and grab the `ID3D12Resource` and pull it out. There are official Vulkan extensions that you can implement in order to do that (`VK_KHR_external_memory`) but this is going to be (much) more effort than just exporting a single function (something non standard) to retrieve the resource directly. But the latter might not be acceptable by the Mesa developers.

**Comment by mbucchia** [comment]
>... exporting a single function (something non standard) to retrieve the resource directly. But the latter might not be acceptable by the Mesa developers.

Something like this in Dozen:

```
ID3D12Resource* getUnderlyingD3D12Resource(VkImage image) {
  return dzn_image_from_handle(image)->res;
}
```

This will then allow you to create a thin interop layer between Dozen and D3D12 for OpenXR, when you would just need to write the code to initialize OpenXR using D3D12, and substitute the D3D12 swapchain images returned by [xrEnumerateSwapchainImages](https://www.khronos.org/registry/OpenXR/specs/1.0/html/xrspec.html#xrEnumerateSwapchainImages) (D3D12 textures) with Vulkan textures, and eventually perform a copy between the underlying Dozen texture (retrieved with `getUnderlyingD3D12Resource()`) into the actual D3D12 swapchain images. Don't forget to also signal a fence to the D3D12 command queue after doing the copy.

**Comment by GeorgeS2019** [comment]
@mbucchia 

The Dozen on WSL is now [working](https://github.com/godotengine/godot-proposals/discussions/4335#discussioncomment-3110853)

**Comment by mbucchia** [comment]
Thanks. I gave some input on what can be implemented, but I don't have any intention to do it myself. I don't have spare cycles for this.

**Comment by rcoppy** [comment]
> 

wow!! thanks so much for this, amazing work 🙂

**Comment by GeorgeS2019** [comment]
@mbucchia @BastiaanOlij 

## [Get Godot Engine to work in Hololens through OpenXR supported by Dozen](https://gitlab.freedesktop.org/mesa/mesa/-/issues/6318#note_1496955)

Please feedback:  from [Jesse Natalie](https://gitlab.freedesktop.org/jenatali)
> [VK_KHR_external_memory_win32](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_external_memory_win32.html) will be implemented and would allow importing/exporting D3D12 resources for interop with Dozen. I don't think it makes sense to implement custom extensions (or non-extension functions) that only Dozen could support at this time.


### Discussion source
![image](https://user-images.githubusercontent.com/49812372/182935644-0dc167f8-fe31-485b-902b-65467f6e4c74.png)

**Comment by GeorgeS2019** [comment]
[Direct3D 12 Rendering Driver ](https://github.com/godotengine/godot/pull/64304/files)

**Comment by BastiaanOlij** [comment]
@GeorgeS2019 cool huh! There are still a few things to be solved but this is making WMR support and Hololens support within reach.

I'm talking with Pedro about the missing bits, I hope to have time in the near future to help him out and see if we can get WMR to work on Windows.

**Comment by GeorgeS2019** [comment]
@BastiaanOlij 

ChatGpt reference this for Hololense and Godot...I assume ..there are better headsets now than Hololens for Godot 4
